pallets/collection-data-feed/src/lib.rs

```diff
@@ -0,0 +1,169 @@
+		type Moment: Parameter + MaxEncodedLen;
+
+		/// Data provider for initializing data values
+		type DataProvider: DataProviderExtended<Self::DataId, (Self::Data, Self::Moment)>;
+
+		/// Max size of a data collection
+		#[pallet::constant]
+		type MaxCollectionSize: Get<u32>;
+
+		/// Max number of collections
+		#[pallet::constant]
+		type MaxCollections: Get<u32>;
+	}
+
+	/// Storage that contains the registering information
+	#[pallet::storage]
+	pub(crate) type Listening<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
+		T::DataId,
+		BoundedBTreeMap<T::CollectionId, u32, T::MaxCollections>,
+		ValueQuery,
+	>;
+
+	/// Storage that contains the data values of a collection.
+	#[pallet::storage]
+	pub(crate) type Collection<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
+		T::CollectionId,
+		BoundedBTreeMap<T::DataId, DataValueOf<T>, T::MaxCollectionSize>,
+		ValueQuery,
+	>;
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The used data ID is not in the collection.
+		DataIdNotInCollection,
+
+		/// Max collection size exceeded
+		MaxCollectionSize,
+
+		/// Max collection number exceeded
+		MaxCollectionNumber,
+	}
+
+	impl<T: Config> DataRegistry for Pallet<T> {
+		type Collection = CachedCollection<T>;
+		type CollectionId = T::CollectionId;
+		type Data = T::Data;
+		type DataId = T::DataId;
+		type Moment = T::Moment;
+
+		fn get(data_id: &T::DataId) -> DataValueOf<T> {
+			T::DataProvider::get_no_op(data_id)
+		}
+
+		fn collection(collection_id: &T::CollectionId) -> Self::Collection {
+			CachedCollection(Collection::<T>::get(collection_id))
+		}
+
+		fn register_data_id(
+			data_id: &T::DataId,
+			collection_id: &T::CollectionId,
+		) -> DispatchResult {
+			Listening::<T>::try_mutate(data_id, |counters| match counters.get_mut(collection_id) {
+				Some(counter) => counter.ensure_add_assign(1).map_err(|e| e.into()),
+				None => {
+					counters
+						.try_insert(collection_id.clone(), 1)
+						.map_err(|_| Error::<T>::MaxCollectionNumber)?;
+
+					Collection::<T>::try_mutate(collection_id, |collection| {
+						collection
+							.try_insert(data_id.clone(), Self::get(data_id))
+							.map(|_| ())
+							.map_err(|_| Error::<T>::MaxCollectionSize.into())
+					})
+				}
+			})
+		}
+
+		fn unregister_data_id(
+			data_id: &T::DataId,
+			collection_id: &T::CollectionId,
+		) -> DispatchResult {
+			Listening::<T>::try_mutate(data_id, |counters| {
+				let counter = counters
+					.get_mut(collection_id)
+					.ok_or(Error::<T>::DataIdNotInCollection)?;
+
+				counter.ensure_sub_assign(1)?;
+				if *counter == 0 {
+					counters.remove(collection_id);
+					Collection::<T>::mutate(collection_id, |collection| collection.remove(data_id));
+				}
+
+				Ok(())
+			})
+		}
+	}
+
+	impl<T: Config> OnNewData<T::AccountId, T::DataId, T::Data> for Pallet<T> {
+		fn on_new_data(_: &T::AccountId, data_id: &T::DataId, _: &T::Data) {
+			// Input Data parameter could not correspond with the data comming from `DataProvider`.
```

Comment:
Why is this so? Can't we restrict the impl of `OnNewData`? This would safe us one storage read right? The one where we fetch from the provider directly instead of using the feeded value.

Reply:
Because from this pallet, you can not ensure that the `DataProvider` and who calls to `on_new_data()` uses the same `Data` value for the same `DataId`.

In fact, the implementation of `orml_oracle` doesn`t use the same. `DataProvider` returns a combined data from several addresses, and when it calls to `on_new_data` uses a non-combined data. See https://github.com/open-web3-stack/open-runtime-module-library/issues/903

Reading only from `DataProvider` we ensure we don't break any invariant.

Reply:
Ah, okay.